File: v1/code/go/src/jvmgo/ch06/instructions/base/instruction.go
package base

import "jvmgo/ch06/rtda"

type Instruction interface {
	FetchOperands(reader *BytecodeReader)
	Execute(frame *rtda.Frame)
}

type NoOperandsInstruction struct {
	// empty
}

func (self *NoOperandsInstruction) FetchOperands(reader *BytecodeReader) {
	// nothing to do
}

type BranchInstruction struct {
	Offset int
}

func (self *BranchInstruction) FetchOperands(reader *BytecodeReader) {
	self.Offset = int(reader.ReadInt16())
}

type Index8Instruction struct {
	Index uint
}

func (self *Index8Instruction) FetchOperands(reader *BytecodeReader) {
	self.Index = uint(reader.ReadUint8())
}

type Index16Instruction struct {
	Index uint
}

func (self *Index16Instruction) FetchOperands(reader *BytecodeReader) {
	self.Index = uint(reader.ReadUint16())
}

File: v1/code/java/jvmgo_java/ch05/src/main/java/com/github/jvmgo/instructions/base/Instruction.java
package com.github.jvmgo.instructions.base;

import com.github.jvmgo.rtda.Frame;

public interface Instruction {


    int getOpCode();

    default String getReName() {
        return this.getClass().getSimpleName();
    }

    /**
     * 从字节码中提取操作数
     *
     * @param reader
     */
    void fetchOperands(BytecodeReader reader) throws Exception;

    /**
     * 执行逻辑指令
     *
     * @param frame
     */
    void execute(Frame frame) throws Exception;
}
